### ISSUE TITLE
Create a \tiifsavedtitleequal togle

### ISSUE THREAD

**Issue by bastien-roucaries** [plk/biblatex#1023]
Hi,

Actually I use this kind of contruct in order to replace identical title between reprint by dash, and I will appreciate a toogle to test if title are identical....
```
\iffieldundef{savedtitleaddon}{%
          \iffieldundef{savedsubtitle}{%
            \iffieldundef{savedtitle}{%
              \iffieldundef{savedmaintitle}{%
                % do nothing
              }%
              {% savedmaintitle defined
                \iffieldsequal{savedmaintitle}{maintitle}{%
                  \restorefield{maintitle}{\dashditto}%
                }{%
                  % do nothing
                }% 
              }%   
            }%
            {% savedtitle defined
              \iffieldundef{savedmaintitle}{%
                % savedmaintitle undefined
                \iffieldsequal{savedtitle}{title}{%
                  \restorefield{title}{\dashditto}%
                }{%
                  % do nothing
                }% 
              }%
              {% savedtitle and savedmaintitle defined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedmaintitle}{maintitle}}%
                }%
                {%
                  \restorefield{maintitle}{\dashditto}%
                  \clearfield{title}%
                }%
                {%
                     % do nothing
                }%
              }%
            }%
          }%
          {% savedsubtitle defined
            \iffieldundef{savedtitle}{%
              % do nothing subtitle without title error likely
            }%
            {% defined savedtitle
              \iffieldundef{savedmaintitle}{%
                % savedmaintitle undefined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedsubtitle}{subtitle}}}%
                {
                  \restorefield{title}{\dashditto}%
                  \clearfield{subtitle}%
                }{%
                  % do nothing
                }% 
              }%
              {% savedtitle and savedmaintitle defined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedmaintitle}{maintitle}}%
                  and test {\iffieldsequal{savedsubtitle}{subtitle}}%
                }%
                {%
                  \restorefield{maintitle}{\dashditto}%
                  \clearfield{title}%
                  \clearfield{subtitle}%
                }%
                {%
                     % do nothing
                }%
              }%
            }%
          }%
        }{% titleaddon defined
          \iffieldundef{savedsubtitle}{%
            % subtitle not defined
            \iffieldundef{savedtitle}{%
              % do nothing subtitle without title error likely
            }%
            {% defined savedtitle, titleaddon, not def subtitle
              \iffieldundef{savedmaintitle}{%
                % savedmaintitle undefined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedtitleaddon}{titleaddon}}%
                }%
                {
                  \restorefield{title}{\dashditto}%
                  \clearfield{titleaddon}%
                }{%
                  % do nothing
                }% 
              }%
              {% savedmaintitle defined savedtitle, titleaddon, not def subtitle
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedmaintitle}{maintitle}}%
                  and test {\iffieldsequal{savedtitleaddon}{titleaddon}}%
                }%
                {%
                  \restorefield{maintitle}{\dashditto}%
                  \clearfield{title}%
                  \clearfield{titleaddon}%
                }%
                {%
                     % do nothing
                }%
              }%
            }
          }%
          {% here titleaddon and subtitle is defined
            \iffieldundef{savedtitle}{%
              % do nothing subtitle without title error likely
            }%
            {% defined savedtitle
              \iffieldundef{savedmaintitle}{%
                % savedmaintitle undefined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedsubtitle}{subtitle}}%
                  and test {\iffieldsequal{savedtitleaddon}{titleaddon}}
                }%
                {
                  \restorefield{title}{\dashditto}%
                  \clearfield{subtitle}%
                  \clearfield{titleaddon}%
                }{%
                  % do nothing
                }% 
              }%
              {% savedtitle and savedmaintitle defined
                \ifboolexpr{ test {\iffieldsequal{savedtitle}{title}}%
                  and test {\iffieldsequal{savedmaintitle}{maintitle}}%
                  and test {\iffieldsequal{savedsubtitle}{subtitle}}%
                  and test {\iffieldsequal{savedtitleaddon}{titleaddon}}
                }%
                {%
                  \restorefield{maintitle}{\dashditto}%
                  \clearfield{title}%
                  \clearfield{subtitle}%
                  \clearfield{titleaddon}%
                }%
                {%
                     % do nothing
                }%
              }%
            }
          }
        }%
```

**Comment by moewew** [comment]
I'm afraid I don't quite understand what the macro you are looking for would do exactly. Obviously something like `\iffieldsequal{savedtitle}{title}` is not enough for you. But what do you want the test to do?

**Comment by bastien-roucaries** [comment]
I want to test if title == savedtitle and (if defined subtitle==savedsubtilte) and (if defined maintitle = savedmaintitle) and if defined (titleaddon==savedtitleaddon)

**Comment by moewew** [comment]
I might be missing something, but here is a test that checks if the fields are equal or both undefined
```latex
% modelled after etoolbox's \ifcsequal
\newcommand*{\blx@ifcsequalorbothundef}[2]{%
  \ifcsundef{#1}
    {\ifcsundef{#2}}
    {\ifcsundef{#2}
       {\@secondoftwo}
       {\expandafter\ifx
        \csname#1\expandafter\endcsname
        \csname#2\endcsname
          \expandafter\@firstoftwo
        \else
          \expandafter\@secondoftwo
        \fi}}}

\def\blx@imc@iffieldequalsorbothundef#1#2{%
  \blx@ifcsequalorbothundef{abx@field@#1}{abx@field@#2}}

\blx@regimcs{\iffieldsequalorbothundef}%  
```
which would be used as `\ifcsequalorbothundef{savedtitleaddon}{titleaddon}{<true>}{<false>}`.

Naturally you can easily collect these tests together for several title types with `\ifboolexpr` - or harder to read but with lazy evaluation
```latex
\newcommand*{\ifsavedtitleequal}{%
  \iffieldsequal{title}{savedtitle}
    {\ifcsequalorbothundef{subtitle}{savedsubtitle}
       {\ifcsequalorbothundef{titleaddon}{savedtitleaddon}}
       {\@secondoftwo}}
    {\@secondoftwo}}
```

**Comment by moewew** [comment]
Urgh, it's too late for me. ... There were quite a few typos in the code above. Sorry about that. Should hopefully be fixed now. I'll get some sleep now.

**Comment by bastien-roucaries** [comment]
@moewew this is the function I needed could you commit this iffieldsequalorbothundef maybe under the name 
iffieldsweakequal ?

**Comment by moewew** [comment]
I'll look into what adding a function like this would entail (we also want to add an equivalent command for lists, names etc.).

@plk Any comments on the name? `\iffieldsequalorbothundef` is a mouthful, so currently I lean towards `\iffieldsequalorundef`, but @bastien-roucaries `\iffieldsweakequal` would also be an option (though I don't know if 'weak equality' is a concept we should try to define).

**Comment by bastien-roucaries** [comment]
@moewew Weak in the mathematical sense. Weak equality

**Comment by bastien-roucaries** [comment]
@moewew if you push the \ifsavedtitleequal could you add a test for maintitle ?

**Comment by moewew** [comment]
I don't think I'll want to add `\ifsavedtitleequal` - it's a bit too specific. But it is easily define, so you are of course free to define something like this (taking into account `maintitle` if you like) in your own style/preamble.

**Comment by moewew** [comment]
Let me know what you think of https://github.com/plk/biblatex/pull/1031

**Comment by moewew** [comment]
As mentioned in in #1031 it is possible to build these kinds of test yourself with what is available in the `biblatex` core at the moment. So these new commands are not strictly speaking required.
E.g.
```latex
\newcommand*{\iffieldequalsorbothundef}[2]{%
  \iffieldequals{#1}{#2}
    {\@firstoftwo}
    {\iffieldundef{#1}
       {\ifundef{#2}}
       {\@secondoftwo}}}
```
technically speaking, these tests are going to be less efficient than the version I suggested in #1031 (and now in https://github.com/plk/biblatex/compare/dev...moewew:equalorundef), but I don't expect that to matter a lot in your average document.

Given the additional maintenance overhead way may not want to include the new tests especially since they can be easily defined yourself (either the straightforward, but slightly less efficient version shown here or the more efficient version using internal commands from #1031).

**Comment by moewew** [comment]
Given that it is already possible to build these commands from what is available in the core right now, it is probably not worth cluttering the default namespace with even more macros. This does not appear to be a hotly-requested macro, so I'm closing this a won't fix and won't be merging #1031.

If it turns out that many people need this and that the DIY solution from above is too slow or scares people away from doing what they want because they have to define a macro that they feel should be in the `biblatex` core, we can reconsider. In that case, please comment here.
